core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -338,7 +340,17 @@ private boolean isInliningCandidate(Expression expression, ProjectNode node)
             return dependencies.entrySet().stream()
                     .allMatch(entry -> entry.getValue() == 1
                             || node.getAssignments().get(entry.getKey()) instanceof Literal
-                            || node.getAssignments().get(entry.getKey()) instanceof SymbolReference);
+                            || node.getAssignments().get(entry.getKey()) instanceof SymbolReference
+                            || isEligibleForLinearExpansion(node.getAssignments().get(entry.getKey()), entry.getValue()));
+        }
+
+        // Relax the heuristics to make common scenarios eligible for predicate pushdown
+        private boolean isEligibleForLinearExpansion(Expression expr, long count)
+        {
+            // Qualify the expression that refers to at most a single symbol
+            // but also avoid excessive duplication of long expression
+            return SymbolsExtractor.extractAll(expr).size() <= 1 &&
+                    expr.toString().length() * count <= MAX_LINEAR_EXPANSION_LENGTH;
```

Comment:
I would rather count some term of complexity instead of string length. Otherwise depending on symbol name you would get different expansion factors.

Reply:
Any suggestion on other criteria? It could be an arbitrary expression here, not sure if there's a generic way to tell the complexity

Reply:
What case are you trying to solve? Maybe we could count addition/subtraction/multipliaction terms?

Reply:
I see some more special handling in that case :) I think "count" should be a pretty good indicator that we can use as part of the limiting heuristics